### DISCUSSION TITLE
next/image hostname error

### DISCUSSION THREAD

**Discussion by gwendall** [vercel/next.js#18311] [Help] [answered]
I get the following error using an image with an absolute url: `Invalid src prop ({REDACTED URL}) on 'next/image', hostname is not configured under images in your 'next.config.js'`.

Nothing in the doc mentions how to configure that. Any hints ?

And congrats on the v10 release !

**Comment by gwendall** [comment, marked as answer]
My bad, just found the answer here: https://github.com/vercel/next.js/blob/canary/examples/image-component/next.config.js

**Reply by RiyaGupta89** [reply]
Not working :(

**Reply by mrlaseptima** [reply]
Not working for me too

**Reply by itsanji** [reply]
Remember to Restart the server MANUALLY!

**Reply by pkstrz** [reply]
@RiyaGupta89 @mrlaseptima Restart the server and make sure you have a valid import. It may sound strange, but in my case it was the problem. I had the correct domains in next.config.js and restarted the server and it still didn't work. Then I realize that I had a bad import. I had:
`import Image from 'next/Image';`
instead of
`import Image from 'next/image';`
And I had this error.

**Reply by Ntagungira-cmd** [reply]
Thanks mn, u just saved a life 👍👍

**Reply by HNazmul-X** [reply]
anybody can tell me please how to allow images from any host in next.js

**Reply by FDiskas** [reply]
Vote for https://github.com/vercel/next.js/pull/27345

**Reply by Shin-Thant** [reply]
My image URL is https://www.freetogame.com/g/1/thumbnail.jpg and how should I add it in domains?

**Reply by hamedmatari** [reply]
@pkostrzewski wow thanks, I had exactly that problem, appreciate for sharing 🙌🏼

**Reply by ventura01** [reply]
Funcionó para mi escribiendo:
next.config.js
`module.exports = {
  images: {
    domains: ['example.com'],
    formats: ['image/avif', 'image/webp'],
  },
}`

**Reply by owaini** [reply]
> anybody can tell me please how to allow images from any host in next.js
you should use this in frontend folder
create next.config.js file in root then write >>
module.exports = {
	images: {
		domains: ["res.cloudinary.com"],  << here your host image
		formats: ["image/avif", "image/webp"],
	},
};

**Reply by ryanp-hrs** [reply]
how do you restart the server manually ?

**Reply by dsope05** [reply]
@ryanp-hrs just exit server with control c, and start server with npm run dev

**Comment by gwendall** [comment]
So other question: can we use images from random remote sources ? Using `images: { domains: ['*'] }` does not seem to work.

**Reply by rokibulcommit** [reply]
even localhost or 127.0.0.1 doesn't seem to work. I have a server running on localhost:5000 and I want to load images from that server

**Reply by rokibulcommit** [reply]
My bad, I had to use only hostname there in the config. If the server is running on port `5000` and host is `localhost` or `127.0.0.1` then only the hostname i.e. `localhost` or `127.0.0.1` to be used.

I'm not deleting my previous reply, it might be helpful to someone who is stuck like me in this same scenario.

**Reply by FDiskas** [reply]
The idea to use wildcard is not bad https://github.com/vercel/next.js/pull/27345

**Reply by HNazmul-X** [reply]
> even localhost or 127.0.0.1 doesn't seem to work. I have a server running on localhost:5000 and I want to load images from that server

so what is the way to allow all images from any host? tell me please if there are anyway.

**Reply by FDiskas** [reply]
There is no way yo allow any host. And it is not recommended for security and service abuse. Anybody could use your service to resize images via url.
But in case you know what you doing vote for pull request it should allow us to specify a regexp for domains if needed https://github.com/vercel/next.js/pull/27345

**Reply by ismailghedamsi** [reply]
@FDiskas what if an app use an api that get images from multiple source. What can it be done. Isn't it a valid use case?

**Reply by FDiskas** [reply]
I do believe it is already implemented. Check my PR last comments.

**Comment by AlexSapoznikov** [comment]
@timneutkens Can we have a bit information on this here please?

**Comment by VivekNeel** [comment]
If the image is svg and present in public/static, how do you load? It's still excepting hostname to be configured!

**Reply by AlexSapoznikov** [reply]
Indeed, confirm that I have the same issue

**Reply by vixtorocha** [reply]
Was able to make it work with **next-compose-plugins**! After installing it, edit **next.config.js**:

```
const withPlugins = require('next-compose-plugins')
const withImages = require('next-images')

const nextConfig = {
  images: {
    domains: ['your.domain']
  }
}

module.exports = withPlugins([[withImages]], nextConfig)
```

Edit: Remember to **restart the server**!

**Reply by peppermint-juli** [reply]
This one worked like a charm, thanks a lot for sharing!

**Reply by johnfreitasau** [reply]
Worked like a charm. Thanks!

**Reply by boakyezitek** [reply]
Bro you have saved me thank you so much

**Reply by seyedkazemmousavi** [reply]
If we have a lot of dynamic URLs what should we do?

**Reply by Arthur-bl-dev** [reply]
> Was able to make it work with **next-compose-plugins**! After installing it, edit **next.config.js**:
> 
> ```
> const withPlugins = require('next-compose-plugins')
> const withImages = require('next-images')
> 
> const nextConfig = {
>   images: {
>     domains: ['your.domain']
>   }
> }
> 
> module.exports = withPlugins([[withImages]], nextConfig)
> ```
> 
> Edit: Remember to **restart the server**!

Worked for me. Thanks!

**Reply by dougmotshell** [reply]
[SOLVED] for me!

@Arthur-bl-dev  No need to install dependencies!!!

I was trying to use github avatar url.

The @miguelseguramx is right!

Just remove the protocol (_https_ or _http_)

`next.config.js`
```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    domains: ['avatars.githubusercontent.com']
  }
}

module.exports = nextConfig
```

`Navbar.tsx`
```tsx
  <Image
    width={48}
    height={48}
    src="https://avatars.githubusercontent.com/u/57239907?v=4"
    objectFit="cover"
    alt="Picture of the author"
    className="h-8 w-8 rounded-full"
/>
```
**Note**: Set the width and height because before I set these properties the image was not showing.

**Reply by leratolsname** [reply]
thank you

**Comment by hatchli** [comment]
Loading an image from a cloudinary url, despite setting the domain as cloudinary.com in next.config.js, doesn't work and triggers this error...

**Reply by marcelobaez** [reply]
Same happens to me with local server (localhost)

**Reply by enochndika** [reply]
With cloudinary you have to use loader, the domain does not support cloudinary yet.
it is resolved here  https://github.com/vercel/next.js/discussions/18374#discussioncomment-114339

**Reply by tsusini** [reply]
@marcelobaez just use `domains: ['localhost']`

**Reply by nisalk96** [reply]
> @marcelobaez just use `domains: ['localhost']`

Yeah that works for me on 'localhost' , thanx

**Reply by dennyzain** [reply]
> @marcelobaez just use `domains: ['localhost']`

that works, thanks mate

**Comment by ijjk** [comment]
Hi, a PR was opened here https://github.com/vercel/next.js/pull/18325 to add an error document explaining where this hostname needs to be added to correct the error and will be available in a patch release soon. Wildcard values are not yet supported such as `*`, although is being investigated for inclusion later down the road.

**Reply by all-the-rice** [reply]
+1 for wildcard

**Comment by 0leksiy** [comment]
I had the same error when wanted to use both cloudinary images and images located in the /public dir. I managed to make it work with the fallowing config:
domains: [‘res.cloudinary.com’]
This allows me to keep the default loader to access assets in the /public dir and if I want to load images from cloudinary I simply pass the absolute url to src prop of the Image component.

**Comment by cyrus-za** [comment]
We use auth0 with social logins. It is a little unfeasible to add each and every potential place a user's profile pic could be stored. For example just google alone requires this:
```ts
 domains: [
      'storage.googleapis.com',
      'lh1.googleusercontent.com',
      'lh2.googleusercontent.com',
      'lh3.googleusercontent.com',
      'lh4.googleusercontent.com',
      'lh5.googleusercontent.com',
      'lh6.googleusercontent.com',
    ],
```

etc. See here for a list of 10k+ sub domains https://securitytrails.com/list/apex_domain/googleusercontent.com

**Reply by seyedkazemmousavi** [reply]
this is good for several URLs

If we have a lot of dynamic URLs what should we do?

**Reply by haukurmar** [reply]
How many of the lhX.googleusercontent.com sub domains are there? 
I see you only list lh1-6.

**Reply by qtv100291** [reply]
Anyone have idea for this ??

**Reply by FDiskas** [reply]
Wildcards domains feature is coming. I seen couple PR already. Wait or contribute :)

**Comment by jackguoAtJogg** [comment]
Anyone have solution for AWS s3 bucket?

**Reply by saintjcob** [reply]
I'm looking for the solution as well.

**Reply by exroot** [reply]
same here

**Reply by miguelseguramx** [reply]
In this case the keyword is domain, _https_, _http_, _www_ its the protocol
The domain it's **something.com** 

Just add your domain in the next.config.js

 ```
 images: {
    domains: ['api-production.s3.amazonaws.com'],
  },
```

**Reply by saintjcob** [reply]
I have tried that. At first I did 'amazonaws.com', then everything that's before 'amazonaws.com', but it still didn't work. I think I'll just keep my pictures local, at least for now and see how it works when the project is finished.

**Reply by foodaka** [reply]
This worked for me for aws s3

```
module.exports = {
    images: {
      domains: ['localhost', 'some-s3-bucket.s3.eu-central-1.amazonaws.com']
    }
  }
```

**Reply by gullerg** [reply]
I have a weird issue where the src of the image references my S3 bucket incorrectly: instead of just having the url of the S3 bucket, /_next/image is attached to the front of the URL. Did you encounter this issue?

**Reply by gullerg** [reply]
Also, if I replace the Image tags with normale img tags and keep the src constant, it works as expected, so it appears to be related to next/image.

**Reply by bhatvikrant** [reply]
> In this case the keyword is domain, _https_, _http_, _www_ its the protocol
> The domain it's **something.com**
> 
> Just add your domain in the next.config.js
> 
> ```
> images: {
>    domains: ['api-production.s3.amazonaws.com'],
>  },
> ```

Thanks for this! It works for amazon s3 bucket

**Reply by Rainase** [reply]
exactly what @bhatvikrant said 👍

**Reply by maguirekrist** [reply]
This does not work. I've tried the next.config.js domains, I haven't been able to find a solution for S3 still. 

edit - It works now, for anyone else reading this, you need to implement the loader function and tag into your Image component. More info here: https://nextjs.org/docs/api-reference/next/image

**Reply by lazarok09** [reply]
on my case i have this on next config 

// next.config.ts
```
module.exports = {
  reactStrictMode: true,
  trailingSlash: true,
  images: {
    domains: ['s3.amazonaws.com'],
  },
};
```

// usage in a normal JSX component - notice that the image don't have https: so i puted it

```
const user_image_image = "//s3.amazonaws.com..."
          <Image
            src={`https:${user_image_image}`}
            alt={` a profile picture`}
            height={'54px'}
            width={'48px'}
          />
```

**Comment by TitusEfferian** [comment]
anyone have solution for firebase storage?
i still got error for this, if my firebase storage url looks like this 
`https://firebasestorage.googleapis.com/v0/b/my_project_name/o/my_file.png`
```
module.exports = {
    images: {
        domains: [
            'https://firebasestorage.googleapis.com/',
          ],
    }
}
```

**Reply by tsusini** [reply]
@TitusEfferian `domains: ['firebasestorage.googleapis.com']`

**Reply by gayathrithedev** [reply]
https://icons8.com/icon/2uKzzjRadgyq/twitter

to configure above domain, I am following below, but it is not working as expected in my case. 

module.exports = {
    images: {
        domains: [ 'icons8.com'],
    }
}

Anyone please help me what is I am missing?

**Reply by ignacioHourquescos** [reply]
had the same issue

> anyone have solution for firebase storage?
> i still got error for this, if my firebase storage url looks like this
> `https://firebasestorage.googleapis.com/v0/b/my_project_name/o/my_file.png`
> 
> ```
> module.exports = {
>     images: {
>         domains: [
>             'https://firebasestorage.googleapis.com/',
>           ],
>     }
> }
> ```

This works fine. must restart server!!!

**Comment by n1kalai** [comment]
when i map array and give image link to components, it caches images and shows them incorrectly :|

**Comment by MalvinJay** [comment]
I've tried everthing most things suggested here but it's not working. Sigh! how is a library that's suppose to make life eary being this troublesome while configuring?

This is my next.config.js

`
const path = require('path');
const withPlugins = require('next-compose-plugins')
const withImages = require('next-images')
  
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})

const nextConfig = {
  images: {
    domains: [
      'localhost',
      'BucketName.s3.us-west-1.amazonaws.com',
      'amazonaws.com',
      'res.cloudinary.com'
    ]
  },
  sassOptions: {
    includePaths: [path.join(__dirname, 'styles')],
  } 
}

module.exports = withPlugins([[withImages]], `nextConfig)`
module.exports = withBundleAnalyzer();
`

And this is the error I keep getting
`
Error: Invalid src prop (https://BucketName.s3.us-west-1.amazonaws.com/images/2oQqtoaESaFaNqPB5YzDt674oRHnje8FCGoGStAKGPcP5krQ.png) on `next/image`, hostname "BucketName.s3.us-west-1.amazonaws.com" is not configured under images in your `next.config.js`
See more info: https://err.sh/next.js/next-image-unconfigured-host
`

**Reply by madzadev** [reply]
also for those editing domains in next.config.js, remember to restart the dev server for changes to take effect.
just a reminder in case you forget, as I did 😀😀

**Reply by edgarasben** [reply]
@madzadev thank you :D

**Reply by findnizam** [reply]
> also for those editing domains in next.config.js, remember to restart the dev server for changes to take effect.
> just a reminder in case you forget, as I did 😀😀

This worked for me.... thanks a lot

**Comment by monecchi** [comment]
I've been using prismic.io to manage a blog, and I was also getting that error when trying to set up `images.prismic.io` domain. After reading next docs about `next-image` component here: https://nextjs.org/docs/api-reference/next/image#loader, the solution which worked for me was using the "loader" prop to resolve the image url...

### next-config.js

``` tsx
module.exports = {
  images: {
    domains: ['images.prismic.io', 'prismic-io.s3.amazonaws.com'],
    // next/image support `srcSet` using the below deviceSizes
    // for more info, visit https://nextjs.org/docs/basic-features/image-optimization#device-sizes
    deviceSizes: [320, 480, 640, 750, 828, 1080, 1200, 1920, 2048, 3840],
    // https://nextjs.org/docs/basic-features/image-optimization#image-sizes
    imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
  },
}
```

### Here's my full implementation

```tsx
import Head from 'next/head'
import { client } from '../../prismic-configuration'
import Prismic from 'prismic-javascript'
import Listing from '../components/Listing'

//
// Blog Main Page
//

const Blog = (props) => {
return (
    <>
      <Head>
        <title>Blog</title>
      </Head>
      <div>
       <Listing articles={props.articles.results} />
      </div>
    </>
}

export async function getStaticProps() {
  const articles = await client.query(
    Prismic.Predicates.at('document.type', 'article'),
    {orderings: '[my.article.created desc]', pageSize: 10, page: 1}
  )
  return {
    props: {
      articles,
    },
  }
}

export default Blog
```

```tsx
//
// Articles Listing Component
//

import PropTypes from 'prop-types'
import {RichText} from 'prismic-reactjs'
import Image from 'next/image'

// data from prismic.io returns the image src as an absolute url, so no need to set up the full url on loader....
const prismicLoader = ({ src, width, quality }) => {
  return `${src}?w=${width}&q=${quality || 75}`
}

const Listing = ({ articles }) => {
  {articles && articles.map((article) => (
    return (
      <>
        <article key={article.uid}>
            <Image
              loader={prismicLoader}
              src={article.data.article_image.url}
              alt={article.data.article_image.alt}
              width={article.data.article_image.dimensions.width}
              height={article.data.article_image.dimensions.height}
            />
            <h5>{RichText.asText(article.data.title)}</h5>
            <p>{RichText.asText(article.data.excerpt)}</p>
        </article>
      </>
    )
  ))}
}

Listing.propTypes = {
  articles: PropTypes.array,
}

export default Listing;

```

**Reply by xaun** [reply]
Same thing for me using AWS S3 after upgrading to 10.1.x from 10.0.x. My next.config.js was correct, but wasn't working until adding the loader function. Thank you

**Reply by julioflima** [reply]
@xaun @monecchi 

Bros this is just a placebo. If you click on the image, the image still is getting from the same server with no optimization. 
BECAUSE:
https://nextjs.org/docs/basic-features/image-optimization#domains

The next documentation is not clear here so in this part it says:
"When loader is set to an external image service, this option is ignored." 

So we consider it is the loader on next.config.js, but it isn't is the parameter on ReactComponent Image `loader`.

#### Solution:
I found the solution for my code, I'm using a lot of configuration so I didn't put this config as the last parameter of withPlugins.
Thanks to this guy I realized:
https://www.tutorialguruji.com/react-js/next-image-configuration-in-next-js-config-file/

**Reply by julioflima** [reply]
@RiyaGupta89

**Reply by monecchi** [reply]
@julioflima Thanks for the tip! The loader thing is indeed working on my code, in this case the images are being pulled from prismic.io servers and I can confirm they are optimized for responsiveness, cropping, quality etc (from their side though), nextjs is really doing nothing regarding optimization here, leading us to the placebo effect… 

On the blog pages I was working on, I also have set the loader when opening the images on a lightbox like solution… I’m not using any plugins on next-config.js because I wanted to keep it clean, but that’s a long time since this discussion started, so I’ll have glimpse on nextjs recent updates and give it a try.

**Comment by SinghalAyushh** [comment]
Just use :  images: {
                domains: [
                    'localhost'
                
                ],
            },


It 'll solve Your issue

**Reply by mammosu** [reply]
`<Image
  unoptimized
  src={imageUrl}
  ...
  />
`
Adding `unoptimized` property then works for me 🚀

**Comment by wellitongervickas** [comment]
Works for me: https://github.com/vercel/next.js/issues/18412#issuecomment-801708220
Create your own image loader maybe help you

**Reply by FDiskas** [reply]
Or hope that nextjs devs will improve this. In my case aws buckets with random names could solve this issue. example pr https://github.com/vercel/next.js/pull/27345

**Comment by itsanji** [comment]
I had the same issue, i go to [this](https://nextjs.org/docs/messages/next-image-unconfigured-host) article which is in Nextjs Docs,
I did the change but nothing happend. 
But then I realise that nextjs said that i **MUST** restart the server when I change next-config.js
so
**Remember to reload the server manually**

**Reply by steven-js** [reply]
Ah, thank you! Restart the server.

**Comment by richardpickett** [comment]
For folks asking for wildcard, here's my quick solution - I create a shell that puts in the image tag.

It doesn't do image optimization, etc, but I can add it if I want to.

One nice thing is it clearly calls out in the code when I'm expecting an external image.

Necessity is the mother of invention.  I too am using google oath, I actually don't want to use a wildecard if it were an option, and I don't want a million domain entries.

components/externalImage.js:
```
import Image from "next/image";

export default function ExternalImage(props) {
  return <img {...props} />;
}
```

Usage:
pages/whatever.js:
```
...
          <ExternalImage src={currentUser.image} />
...
```

**Reply by adrianhorning08** [reply]
I just used <img /> and ignored the linter and everything seems fine 🤷‍♂️

**Comment by markcnunes** [comment]
Be sure you are running your tests with `NODE_ENV=test` e.g. `"test": "NODE_ENV=test jest"` if you are only getting this error while running tests.

https://github.com/vercel/next.js/pull/26502

**Reply by Tosinkoa** [reply]
I'm getting the error while running npm run dev

**Comment by OZZlE** [comment]
next.config.mjs:

    /**
     * @type {import('next').NextConfig}
     */
    const nextConfig = {
        trailingSlash: true,
        images: {
            domains: ['localhost', 'example.com'],
            formats: ['image/webp', 'image/jpeg', 'image/png', 'image/gif'],
        },
    }

    export default nextConfig

"Error: Specified images.formats must be length 1 or 2, received length (4), please reduce the length of the array to continue."

.. and default eslint config doesn't allow me to use `<img>` ...

It works if you just ommit "formats" but then what's the point with that then?

**Reply by Tosinkoa** [reply]
Got it solved, thanks, I use the specify the url like this: abcdefgh.com
I also specify the width and height.

**Comment by Oskku** [comment]
https://nextjs.org/docs/messages/next-image-unconfigured-host here is the answer

**Comment by kornjiradej** [comment]
i just remove the prefix url 👍 work for me!!

`images: {
 domains: [ 'https://www.google.com' ]
}`

to
`images: {
 domains: [ 'google.com' ]
}`

**Comment by dougmotshell** [comment]
[SOLVED] for me!

@Arthur-bl-dev  No need to install dependencies!!!

I was trying to use github avatar url.

The @miguelseguramx is right!

Just remove the protocol (_https_ or _http_)

`next.config.js`
```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    domains: ['avatars.githubusercontent.com']
  }
}

module.exports = nextConfig
```

`Navbar.tsx`
```tsx
  <Image
    width={48}
    height={48}
    src="https://avatars.githubusercontent.com/u/57239907?v=4"
    objectFit="cover"
    alt="Picture of the author"
    className="h-8 w-8 rounded-full"
/>
```
**Note**: Set the width and height because before I set these properties the image was not showing.

**Comment by umutbozdag** [comment]
I had to restart my server after I made changes to `next.config.js` because it didn't apply my newly added hostname.

**Comment by HNazmul-X** [comment]
why is this happening? do you know anything about this?

**Comment by AvivBenY** [comment]
this works for all domains

images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "**",
      },
    ],
  }

**Reply by hmthu30** [reply]
images: {
		remotePatterns: [
			{
				protocol: "https",
				hostname: imgFrontendDomain,
				port: "",
				pathname: "**",
			},
			{
				protocol: "https",
				hostname: imgStrapiDomain,
				port: "",
				pathname: "**",
			},
		],
	}, i want to dynamically fetch from .env how can i do?

**Reply by lazarok09** [reply]
declare it in the .env file and at hostname put process.env.ENV_NAME. Remember to add those same variables at your infra

**Reply by hmthu30** [reply]
doing with that flow but fail to deploy in UAT server 
@lazarok09

**Reply by lazarok09** [reply]
Check if in your UAT server the .env file is available and have the variables you puted at your local enviroment

**Comment by sneko** [comment]
Despite the settings mentioned in other answers (I tested new ones `images. remotePatterns` and legacy ones `images.domains`), none was working.

After upgrading from `v13.0.7` to `v13.1.6` it works now :)

**Comment by cdpark0530** [comment]
Can anyone explain why Next.js decided to restrict referencing external images?

**Reply by timneutkens** [reply]
Because otherwise anyone can hijack your image optimization endpoint to resize images and you'd pay for it. The default is secure but you're not restricted in any way to allow "any url" using [remotePatterns](https://beta.nextjs.org/docs/api-reference/components/image#remotepatterns) but in practice that is a risk. Wouldn't recommend it.

**Comment by bloodykheeng** [comment]
just add this in ur next.config.js

```
module.exports = {
  images: {
    domains: ['127.0.0.1','localhost'], // 👈 Add this line
  },
}
```
